Visual Studio compatibility fixes

desowin opened the pull request:
The generated C source file is portable. This is preliminary change to
compile libopenvizsla with Visual Studio.

desowin committed:
Convert ov3.fwpkg to C source file

The generated C source file is portable. This is preliminary change to
compile libopenvizsla with Visual Studio.

desowin commented:
The ultimate goal is to get libopenvizsla packaged in [vcpkg](https://github.com/microsoft/vcpkg/issues/8033). Having vcpkg port for libopenvizsla would make it easy to compile on Windows.

matwey commented:
Hi,

I am very glad to see your contribution here. Please, allow me some time to review it. I'll have the next week busy.

desowin committed:
Make code compatibile with Visual Studio

Use uint8_t *data in struct bit instead of void pointer as arithmetic
operations are not valid on void pointers (void pointer arithmetics is
non portable gcc extension).

Use #pragma pack(push, 1) and #pragma pack(pop) instead of
__attribute__((packed)) when building with Visual Studio.

Do not include arpa/inet.h as it is not available on Windows. Provide
simple uint16_from_big_endian() and uint32_from_big_endian() inline
functions to be used instead.

desowin commented:
I don't know why the status after programming differs when I run the code on Linux and Windows. What I call "status" is what gets received on Interface B after sending 0x83 on it (after sending the fwpkg).

On Linux, the three subsequent status reads on interface B yield 0xB8, 0xB8 and finally 0xBC.
On Windows, the three subsequent status reads on interface B yield 0x98, 0x98 and then 0x98.

Do you know what the 0x98 means?

desowin commented:
The only differences I spotted in Windows are before the libftdi initializes device:
- A bit different control packets order during FTDI enumeration
- Windows tries to get MS Feature Descriptors
- Windows calls GET STATUS twice after reading the STRING descriptors

Once the libftdi kicks in with Reset on Port A, the payloads ~~match between Linux and Windows (when comparing usbmon and USBPcap capture) up to the status read stage (0xB8 vs 0x98 as mentioned in previous commit)~~ don't match due to the race condition in send buffer.

This leads me to believe it is some kind of timing issue, but I didn't yet grasp the OV schematic, Xilinx and ~~FTDI datasheets~~ up to the point where it is possible to understand the meaning of the data.

The data on Interface B is simply FTDI MPSSE. It is described in Application Note AN_108 "Command Processor for MPSSE and MCU Host Bus Emulation Modes", so the status byte is just the state of the pins.

desowin commented:
It is a timing problem. There is a race condition in accessing the buf in bit_load_firmware(). Solving the race condition, results in fully working firmware load on Windows.

desowin commented on src/bit.c:
```diff
@@ -159,16 +166,18 @@ int bit_load_firmware(struct bit* bit, struct cha* cha, struct chb* chb) {
 		}
 
-		if (!(tc = ftdi_write_data_submit(&cha->ftdi, buf, i))) {
+		if (!(tc = ftdi_write_data_submit(&cha->ftdi, buf[bidx], i))) {
```
Alternative solution is to simply change the call to ftdi_write_data_submit() with ftdi_write_data()

desowin replied:
Alternative solution is to simply change the call to ftdi_write_data_submit() with ftdi_write_data().

desowin commented:
With the delays the code is working reliably on my OpenVizsla board, both on Windows and Linux.

desowin commented:
The initial extcap works fine on Linux. However on Windows it fails pretty quickly (after few houndred packets) with "Wrong packet magic".

desowin commented:
The problem with "Wrong packet magic" is caused by read buffer race condition in cha_loop_transfer_callback(). When the race condition is fixed, it works as reliably on Windows as it does on Linux.

matwey commented:
Hello.

Thank you for your great contributions. Could I ask you to create separate pull-request per issue/feature? I see that you found a lot of bugs. This would be simpler for me to discuss.

matwey commented:
Frankly speaking, I don't like idea of converting ov3.fwpkg to a source code file much. Why can't we put it in DLL as a resource in Windows? What do you think?

desowin commented:
> Frankly speaking, I don't like idea of converting ov3.fwpkg to a source code file much. Why can't we put it in DLL as a resource in Windows? What do you think?

Using resource files on Windows feels like way too much hassle to me. The biggest disadvantage is that it'd be a Windows-specific solution. In my opinion the proper way to handle multi-platform is to favor portable solutions over a handful of target-specific ones.

The conversion to .c/.h file pair is a simple and portable way to achieve the goal of having arbitrary binary file contents available in C array. This will work on any platform that does have C compiler. Such approach is really common, e.g. gdk-pixbuf-csource or bin2c does just that. Here, no new helper tools are introduced as CMake is capable of achieving the goal.

desowin commented:
> Thank you for your great contributions. Could I ask you to create separate pull-request per issue/feature? I see that you found a lot of bugs. This would be simpler for me to discuss.

There is the option to display individual commits in "Files changed" tab. If that doesn't work well, what about splitting it into 3 pull requests:
- bug fixes
- extcap
- Visual Studio compatibility

matwey commented:
> > Thank you for your great contributions. Could I ask you to create separate pull-request per issue/feature? I see that you found a lot of bugs. This would be simpler for me to discuss.
> 
> There is the option to display individual commits in "Files changed" tab. If that doesn't work well, what about splitting it into 3 pull requests:
> 
>     * bug fixes
> 
>     * extcap
> 
>     * Visual Studio compatibility

Yep, something like that would be fine, Thank you.

desowin commented:
Bug fixes are in #9 and extcap is in #10. I'll put the VS changes after these are reviewed, corrected and merged.

matwey commented:
I have a question on Visual Studio. As far as I can see, `CMakeLists.txt` have to be adapted to be able to used with VS.

The major part is that currently `CMakeLists.txt` relies on `pkg-config` tool to find dependencies. Unfortunately, only LibFTDI1 has native `CMake` module support. Does `vcpkg` has some kind of `pkg-config` emulation or so?

desowin commented:
> The major part is that currently `CMakeLists.txt` relies on `pkg-config` tool to find dependencies. Unfortunately, only LibFTDI1 has native `CMake` module support. Does `vcpkg` has some kind of `pkg-config` emulation or so?

No. `vcpkg` does not support `pkg-config`. If you check the `vcpkg` ports, they basically patch the `CMakeLists.txt` that require `pkg-config` to not require it. I didn't manage to cleanly generate Visual Studio solution - to get the code to compile and link I have manually modified the `.sln` file that I generated using `cmake` to contain required preprocessor and linker config.

matwey commented:
It means that we have to write three CMake modules for:
- [ ] check
- [ ] libusb
- [ ] libzip

desowin commented:
The libftdi1 does pull in libusb (however, there's bug in https://github.com/microsoft/vcpkg/issues/8550)

For the `ovextcap` to work, I had to also add `getopt.lib` to linker options.

desowin commented:
This is made obsolete by #9, #10, #11 and #12.

matwey commented:
@desowin do I understand correctly that you also doesn't run unit tests on Visual Studio? It seems that `libcheck` is missed or broken in `vcpkg`,

desowin commented:
> @desowin do I understand correctly that you also doesn't run unit tests on Visual Studio? It seems that `libcheck` is missed or broken in `vcpkg`,

No, I have managed to run the unit tests just fine with Visual Studio. It seems to work with package `check` version `0.12.0-2` from `vcpkg`.

```
C:\Users\desow>C:\Development\libopenvizsla\RelWithDebInfo\test_bit.exe
Running suite(s): bit
100%: Checks: 2, Failures: 0, Errors: 0

C:\Users\desow>C:\Development\libopenvizsla\RelWithDebInfo\test_decoder.exe
Running suite(s): decoder
100%: Checks: 7, Failures: 0, Errors: 0

C:\Users\desow>C:\Development\libopenvizsla\RelWithDebInfo\test_fwpkg.exe
Running suite(s): fwpkg
100%: Checks: 7, Failures: 0, Errors: 0

C:\Users\desow>C:\Development\libopenvizsla\RelWithDebInfo\test_reg.exe
Running suite(s): reg
100%: Checks: 4, Failures: 0, Errors: 0
```

```
$ ls /c/Development/vcpkg/installed/vcpkg/info
bzip2_1.0.6-4_x64-windows.list
bzip2_1.0.6-4_x86-windows.list
check_0.12.0-2_x64-windows.list
check_0.12.0-2_x86-windows.list
getopt_0_x86-windows.list
getopt-win32_0.1_x86-windows.list
libftdi1_1.4_x64-windows.list
libftdi1_1.4_x86-windows.list
libusb_1.0.22-4_x64-windows.list
libusb_1.0.22-4_x86-windows.list
libzip_rel-1-5-2_x64-windows.list
libzip_rel-1-5-2_x86-windows.list
openssl_1_x64-windows.list
openssl_1_x86-windows.list
openssl-windows_1.0.2s-1_x64-windows.list
openssl-windows_1.0.2s-1_x86-windows.list
zlib_1.2.11-5_x64-windows.list
zlib_1.2.11-5_x86-windows.list
```
